### ISSUE TITLE
Injector Access

### ISSUE THREAD

**Issue by gte619n** [gilberto-torrezan/gwt-views#8]
Hi Gilberto, I've got a situation using gwt-views and I'm wondering if you could comment. The short version is that I think I need access to the Ginjector through a static method, but here's the situation:
- My site uses cookies to persist a device Id and token. 
- When the user returns to the site, these are sent to the server to exchange for a user.
- The user is returned to the client and accessible throughout the application in the UserPresenceManager.

Previously, I created a static injector in my EntryPoint and did this stuff there. I'd like to follow the same pattern, but I can't seem to access or set the injector that is used to manage set the components of the views.

Do you have any ideas about I could accomplish a similar thing and still be within the framework?

Thanks!

E

**Comment by gilberto-torrezan** [comment]
Hi, thanks for your interest in gwt-views =)

Well, to define the injector that will manage the view, you can use the [injector](https://github.com/gilberto-torrezan/gwt-views#dependency-injection) parameter at the `@View` annotation:

``` java
@View(value = "injected", injector = MyInjector.class, injectorMethod = "getInjectedView")
public class InjectedView extends Composite {
```

Inside your view, you can access the injector by using:

``` java
@View(value = "injected", injector = MyInjector.class, injectorMethod = "getInjectedView")
public class InjectedView extends Composite {
    private final MyInjector injector = GWT.create(MyInjector.class);
```

I don't know if just that solves your problem or there's something else.

**Comment by gte619n** [comment]
Hey Gilberto,

Thanks for the response. I need to access the Ginjector in the EntryPoint and then have that Ginjector be the same one that operates the views. The userProvider is a singleton that has the user stored on the client. 

Right now, if I create a ginjector in the EntryPoint using GWT.create, I get a SECOND instance of the ginjector in the views. Is there any way to access the ginjector instance in another way?

Thanks,

E

**Comment by gilberto-torrezan** [comment]
Maybe you can inject the injector to get the same instance:

``` java
public class MyView extends Composite {
    @Inject Injector injector;
```

Found that [on this answer](http://stackoverflow.com/questions/2176216/how-to-inject-injector) on StackOverflow. I don't know if it works with GWT.

**Comment by gte619n** [comment]
Hey Gilberto,

I think the issue might stem from how gwt-views handles the injector
creation. I haven't looked at the code super closely, but it looks like in
PresenterGenerator, line 122, that it's creating a new injector for each
View? If so, I think that's going to break dependency injection, at least
for anything that has share state.

Am I reading this correctly?

Thanks,

E

On Mon, Nov 30, 2015 at 5:38 PM, Gilberto Torrezan <notifications@github.com

**Comment by gilberto-torrezan** [comment]
Exactly. The purpose here was not to "create a new injector for each View"  but to use the injector referenced by the `GWT.create` (as stated on the [GIN tutorial](https://code.google.com/p/google-gin/wiki/GinTutorial#Step_5._Creating_the_Ginjector)). In theory this shouldn't create a new injector each time is called. If you are sure that's not the case, then another approach should be implemented.

**Comment by gte619n** [comment]
Hey Gilberto,

So in all my previous implementations and all of my testing in this case, I've only ever created one static injector. Calling GWT.create( Injector.class ) creates a new injector, and an entirely new stack of objects each time. I've attached a simple test case to show that it's creating new ones.

I believe this is the intended behavior of GIN. If you're within the injection scope through direct traversal, it will provide the same injector and references, but subsequent GWT.Create will make new scopes.

[TestClientModule.java.txt](https://github.com/gilberto-torrezan/gwt-views/files/48535/TestClientModule.java.txt)
[TestGinEntryPoint.java.txt](https://github.com/gilberto-torrezan/gwt-views/files/48536/TestGinEntryPoint.java.txt)
[TestGinjector.java.txt](https://github.com/gilberto-torrezan/gwt-views/files/48537/TestGinjector.java.txt)

**Comment by gilberto-torrezan** [comment]
Ok, thanks for your comments so far. I'll make a way to be possible to set the injector used by the views at the NavigationManager as soon as possible.

**Comment by gilberto-torrezan** [comment]
I just released the v.1.3.1. You can now set the injector instance to be used by the Views at the NavigationManager.

**Comment by gte619n** [comment]
Hey Gilberto,

I noticed the method. Thanks for the quick turnaround on that! I really appreciate it. When I switched over and ran my class, I think the generator might have an error in it. In my LandingPagePlace_LandingPagePlacePresenterImpl class (the generated one), it looks like it is still looking for a Widget on the first line. The line is this:

```
Widget com.hs.auto.app.client.conf.ApplicationGinjector injector = NavigationManager.getInjectorInstance(com.hs.auto.app.client.conf.ApplicationGinjector.class);
```

When it should not have the Widget type in front of it. I think the Widget class is wrong?

Thanks again,

E
